osm2streets-py/src/lib.rs

```diff
@@ -0,0 +1,454 @@
+        } else {
+            street_network.apply_transformations(transformations, &mut timer);
+        }
+
+        Ok(Self {
+            inner: street_network,
+            ways: doc.ways,
+        })
+    }
+
+    /// Converts the entire `StreetNetwork` to a GeoJSON format.
+    ///
+    /// Returns a GeoJSON string representing all elements in the street network.
+    pub fn to_geojson_plain(&self) -> PyResult<String> {
+        self.inner
+            .to_geojson(&Filter::All)
+            .map_err(|e| PyErr::new::<pyo3::exceptions::PyRuntimeError, _>(format!("{}", e)))
+    }
+
+    /// Converts lane polygons in the `StreetNetwork` to a GeoJSON format.
+    ///
+    /// Returns a GeoJSON string representing the polygons of each lane.
+    pub fn to_lane_polygons_geojson(&self) -> PyResult<String> {
+        self.inner
+            .to_lane_polygons_geojson(&Filter::All)
+            .map_err(|e| PyErr::new::<pyo3::exceptions::PyRuntimeError, _>(format!("{}", e)))
+    }
+
+    /// Converts lane markings in the `StreetNetwork` to a GeoJSON format.
+    ///
+    /// Returns a GeoJSON string representing the lane markings, such as dashed or solid lines.
+    pub fn to_lane_markings_geojson(&self) -> PyResult<String> {
+        self.inner
+            .to_lane_markings_geojson(&Filter::All)
+            .map_err(|e| PyErr::new::<pyo3::exceptions::PyRuntimeError, _>(format!("{}", e)))
+    }
+
+    /// Converts intersection markings in the `StreetNetwork` to a GeoJSON format.
+    ///
+    /// Returns a GeoJSON string representing the markings at intersections.
+    pub fn to_intersection_markings_geojson(&self) -> PyResult<String> {
+        self.inner
+            .to_intersection_markings_geojson(&Filter::All)
+            .map_err(|e| PyErr::new::<pyo3::exceptions::PyRuntimeError, _>(format!("{}", e)))
+    }
+
+    /// Retrieves debugging steps for each modification applied to the `StreetNetwork`.
+    ///
+    /// Returns a vector of `PyDebugStreets` objects, which represent intermediate states or steps
+    /// in the network transformation process.
+    pub fn get_debug_steps(&self) -> PyResult<Vec<PyObject>> {
+        Python::with_gil(|py| {
+            Ok(self
+                .inner
+                .debug_steps
+                .iter()
+                .map(|x| PyDebugStreets { inner: x.clone() }.into_py(py))
+                .collect())
+        })
+    }
+
+    /// Converts the clockwise ordering debug information to a GeoJSON format.
+    ///
+    /// Returns a GeoJSON string showing the ordering of intersections in a clockwise manner, which
+    /// can be helpful for debugging road connections.
+    pub fn debug_clockwise_ordering_geojson(&self) -> PyResult<String> {
+        self.inner
+            .debug_clockwise_ordering_geojson(&Filter::All)
+            .map_err(|e| PyErr::new::<pyo3::exceptions::PyRuntimeError, _>(format!("{}", e)))
```

Comment:
Could we define a helper function like `err_to_py` and then call `map_err(err_to_py)` in most places in this file? Maybe two different methods, since I see `PyRuntimeError` and `PyValueError` used. I haven't played around with it, but the equivalent on the JS side looks like this:

```
fn err_to_js<E: std::fmt::Display>(err: E) -> JsValue {
    JsValue::from_str(&err.to_string())
}
```

The generic error type is just anything that can be `to_string()`'ed.

Reply:
Good point, it’s much clearer now. I’ve added the respective functions and included explanations for why both PyRuntimeError and PyValueError are handled differently:

- PyRuntimeError
Used for unexpected runtime failures.
- PyValueError
Used for invalid inputs or incorrect arguments passed by the user.

This has been addressed in commit deea1d1 - Feel free to check the [diff here](https://github.com/lukasbeuster/osm2streets/commit/deea1d1418c2ed20cf3c787cbb5223e8bb3f089f)